### PR TITLE
ci: Refactor the workflows

### LINKED ISSUE #296: Add notifications for nightly build failures
## Description
Nightly builds can fail silently without notifying the team, we should add a way either through slack or email integration to let the team know.

Example failure: https://github.com/kedro-org/kedro-plugins/actions/runs/5803202341
Possible solution: https://github.com/slackapi/slack-github-action

## Context
Why is this change important to you? How would you use it? How can it benefit other users?

### LINKED ISSUE #316: Refactor GitHub Actions CI setup
## Description
Adopt reusable workflows like how we do with the main kedro repo: Split lint, unit tests, and e2e tests into separate, reusable workflows. These can then be called by other workflows (like `kedro-docker/airflow/datasets/telemetry.yml`) with specific parameters such as plugin, python versions, and machines.

## Context
After talking with @ankatiyar we decided it would be best to avoid using a setup-matrix job in `check-plugin.yml`. We can potentially pass the python version matrix and runs-on attributes as inputs directly from kedro-docker.yml, kedro-datasets.yml, and others instead of using a separate job.

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

* Refactor the CI, break down `check-plugin.yml` into reusable workflows that are individually called by each plugin. Fix #316 
* Add notification for nightly build failure. Fix #296 

### Reusable workflows: `unit-tests.yml`, `lint.yml`, `e2e-tests.yml` -
Broke down `check-plugin.yml` which had a lot of case based jobs into reusable workflows -
* `unit-test.yml`
* `lint.yml`
* `e2e-tests.yml`
These workflows receive the `python-version`, `os`, and `plugin` as inputs from caller workflows.

### Caller workflows: `kedro-airflow.yml`, `kedro-datasets.yml`, `kedro-docker.yml`, `kedro-telemetry.yml`
These workflows call the individual reusable workflows with the `plugin`, `os`, and `python-version` depending on which plugin needs which tests. 
* `kedro-airflow` : Calls `unit-tests`, `lint`, and `e2e-tests`.
* `kedro-datasets`: Calls `unit-tests`, `lint`, and `RTD-build`(which is not a reusable workflow since it's only used by datasets)
* `kedro-docker`: Calls `unit-tests`, `lint`, and `e2e-tests`.
* `kedro-telemetry`: Calls `unit-tests`, `lint`, and `e2e-tests`.

NOTE: I've also made these workflows reusable so that all tests for a plugin can be run by simply invoking these workflows, e.g. in `nightly-build.yml` and `check-release.yml`

### Release workflow: `check-release.yml`
Mostly left unchanged but - 
* Instead of a singular `test` job which is needed for the `build-publish` job, it now has `airflow-test`, `datasets-test`, `docker-test`, and `telemetry-test` which runs the whole test suite for the plugin that is being released. 
* As long as all the tests pass or are skipped, it proceeds as usual

NOTE: Tested on my fork by bumping telemetry version and removing the actual github release and PyPI publishing steps from the workflow. See the [workflow file](https://github.com/ankatiyar/kedro-plugins/actions/runs/6072374550/workflow) and [the run](https://github.com/ankatiyar/kedro-plugins/actions/runs/6072374550/workflow).

### Nightly build and notification: `nightly-build.yml`
This workflow also calls the reusable composite workflows `kedro-airflow/datasets/docker/telemetry.yml`. Runs everyday at midnight UTC.
For each plugin -
* `<plugin>-test` is called. 
* The second job `notify-plugin` runs incase of a failure and **creates a new issue** with the label `<plugin> nightly build` if it **doesn't exist/closed previously**, or **adds a comment to a pre-existing issue if it does**. 

NOTE: This creation of issue is based on the label not the issue heading.
NOTE: Tested on my fork by deliberately failing tests. See the created issues here - https://github.com/ankatiyar/kedro-plugins/issues

## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: .github/workflows/nightly-build.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  datasets-test:
+    uses: ./.github/workflows/kedro-datasets.yml
+
+  docker-test:
+    uses: ./.github/workflows/kedro-docker.yml
+
+  telemetry-test:
+    uses: ./.github/workflows/kedro-telemetry.yml
+
+  notify-airflow:
```
Comment (merelcht) [inline .github/workflows/nightly-build.yml:22]: Nice! 👍

**Review by merelcht — APPROVED** [review]
This looks great, thanks @ankatiyar ! Awesome solution for the build failure notification through ticket creation 👍 Great description on the PR again as well ⭐  

Does this help at all with https://github.com/kedro-org/kedro-plugins/issues/218? cc: @SajidAlamQB

**Comment by ankatiyar** [comment]
It doesn't help with #218, for that we'd need to change the `check-version` job.
We could look into splitting the release workflows for each plugin too in the future.

**Review by SajidAlamQB — APPROVED** [review]
This is some heroic work @ankatiyar! 🌟 The CI looks so much neater thank you for taking on this on and killing many birds with one stone!

**Comment by SajidAlamQB** [comment]
> Does this help at all with #218? cc: @SajidAlamQB

Not quite @merelcht we will have to change the `github_actions_release.py` script as that only triggers one release at a time.

File: .github/workflows/check-release.yml [hunk 2]
```diff
@@ -31,16 +31,33 @@ jobs:
...
+    needs: check-version
+    if: ${{ needs.check-version.outputs.new_release == 'true'  &&  needs.check-version.outputs.package_name == 'kedro-telemetry' }}
+    uses: ./.github/workflows/kedro-telemetry.yml
 
   build-publish:
-    needs: [check-version, test]
-    if: ${{ needs.check-version.outputs.new_release == 'true' }}
+    needs: [ check-version, airflow-test, docker-test, datasets-test, telemetry-test ]
+    if: |
+      always() &&
```
[the remark this answers is not in the document]
Comment (ankatiyar) [reply .github/workflows/check-release.yml:57]: This solution is from here -> https://github.com/actions/runner/issues/491#issuecomment-850884422
The build-publish job will still run if any of the needed tests are skipped (which they are if they're not due for a release)

**Pull request merged**